### PR TITLE
[Fixes #43] Rename ErrorHandler and its associated types/extensions

### LINKED ISSUE #43: Rename ErrorHandler and its associated types/extensions
As stated here by @Tratcher (https://github.com/aspnet/Security/issues/55#issuecomment-59231290), the "ErrorHandler is just for unhandled exceptions". This general/generic name is terribly confusing, as it gives you the impression that it's meant to handle every kind of errors and not just uncaught exceptions.

`ExceptionHandler`, `UnhandledExceptionHandler` or `UncaughtExceptionHandler` might be better names.

### PR DESCRIPTION
**Pull request by kichalla**
@Eilon @rynowak

### CODE REVIEW

File: samples/ErrorHandlerSample/Startup.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public void Configure(IApplicationBuilder app)
                     await context.Response.WriteAsync("<html><body>\r\n");
                     await context.Response.WriteAsync("We're sorry, we encountered an un-expected issue with your application.<br>\r\n");
 
-                    var error = context.Features.Get<IErrorHandlerFeature>();
+                    var error = context.Features.Get<IExceptionHandlerFeature>();
```
Comment (kichalla) [inline samples/ErrorHandlerSample/Startup.cs]: This rename is fine as I noticed that this type isn't being used any where in aspnet org.

File: src/Microsoft.AspNet.Diagnostics/Constants.cs [hunk 2]
```diff
@@ -1,11 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-namespace Microsoft.AspNet.Diagnostics
-{
-    internal static class Constants
```
Comment (kichalla) [inline src/Microsoft.AspNet.Diagnostics/Constants.cs]: dead code

File: src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs [hunk 3]
```diff
@@ -31,7 +31,30 @@ public static IApplicationBuilder UseErrorPage([NotNull] this IApplicationBuilde
...
+            return builder.UseMiddleware<DeveloperExceptionPageMiddleware>(options);
+        }
+
+        /// <summary>
+        /// Captures synchronous and asynchronous exceptions from the pipeline and generates HTML error responses.
+        /// Full error details are only displayed by default if 'host.AppMode' is set to 'development' in the IApplicationBuilder.Properties.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public static IApplicationBuilder UseDeveloperExceptionPage([NotNull] this IApplicationBuilder builder)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs]: I am following the approach @Eilon suggested. Adding these methods in addition to existing ones. Will remove them once I finish cross repo changes and check them in.

File: src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs [hunk 4]
```diff
@@ -31,7 +31,30 @@ public static IApplicationBuilder UseErrorPage([NotNull] this IApplicationBuilde
         /// <returns></returns>
         public static IApplicationBuilder UseErrorPage([NotNull] this IApplicationBuilder builder, ErrorPageOptions options)
         {
-            return builder.UseMiddleware<ErrorPageMiddleware>(options);
+            return builder.UseMiddleware<DeveloperExceptionPageMiddleware>(options);
+        }
+
+        /// <summary>
+        /// Captures synchronous and asynchronous exceptions from the pipeline and generates HTML error responses.
+        /// Full error details are only displayed by default if 'host.AppMode' is set to 'development' in the IApplicationBuilder.Properties.
```
Comment (kichalla) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs]: refers to dead code, probably will remove it.

File: src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerOptions.cs [hunk 5]
```diff
@@ -6,10 +6,10 @@
 
 namespace Microsoft.AspNet.Diagnostics
 {
-    public class ErrorHandlerOptions
+    public class ExceptionHandlerOptions
```
Comment (kichalla) [inline src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerOptions.cs]: Safe to rename as this type is not being used in any other repos in aspnet org.

File: src/PageGenerator/Program.cs [hunk 6]
```diff
@@ -35,20 +35,31 @@ public void Main(string[] args)
...
 
             var fileCount = 0;
-            foreach (var fileName in cshtmlFiles)
+            foreach (var viewDir in viewDirectories)
             {
-                Console.WriteLine("  Generating code file for view {0}...", Path.GetFileName(fileName));
-                GenerateCodeFile(fileName, string.Format("{0}.Views", args[0]));
-                Console.WriteLine("      Done!");
-                fileCount++;
+                Console.WriteLine();
```
Comment (kichalla) [inline src/PageGenerator/Program.cs]: These changes were required as I rearranged the (*.cshtml) files in the project to be present in their own respective folders. After the changes:

image

File: samples/ErrorHandlerSample/Startup.cs [hunk 7]
```diff
@@ -5,14 +5,14 @@
 using Microsoft.AspNet.Http.Features;
```
Comment (Tratcher) [inline samples/ErrorHandlerSample/Startup.cs]: Rename the directory.

File: samples/ErrorPageSample/Startup.cs [hunk 8]
```diff
@@ -1,19 +1,19 @@
 using System;
 using Microsoft.AspNet.Builder;
```
Comment (Tratcher) [inline samples/ErrorPageSample/Startup.cs]: Rename the dir.

File: src/Microsoft.AspNet.Diagnostics.Abstractions/IExceptionHandlerFeature.cs [hunk 9]
```diff
@@ -5,7 +5,7 @@
 
 namespace Microsoft.AspNet.Diagnostics
 {
-    public interface IErrorHandlerFeature
+    public interface IExceptionHandlerFeature
     {
         Exception Error { get; }
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Diagnostics.Abstractions/IExceptionHandlerFeature.cs]: Should the property be renamed too?

File: src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        /// The request will not be re-executed if the response has already started.
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="configure"></param>
+        /// <returns></returns>
+        public static IApplicationBuilder UseExceptionHandler(this IApplicationBuilder app, Action<IApplicationBuilder> configure)
+        {
+            var subAppBuilder = app.New();
+            configure(subAppBuilder);
+            var errorPipeline = subAppBuilder.Build();
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerExtensions.cs]: exceptionHandlerPipeline

File: src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerFeature.cs [hunk 11]
```diff
@@ -5,7 +5,7 @@
 
 namespace Microsoft.AspNet.Diagnostics
 {
-    public class ErrorHandlerFeature : IErrorHandlerFeature
+    public class ExceptionHandlerFeature : IExceptionHandlerFeature
     {
         public Exception Error { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Diagnostics/ExceptionHandler/ExceptionHandlerFeature.cs]: Rename the property?

**Comment by Tratcher** [comment]
:watch:

File: samples/ErrorHandlerSample/Startup.cs (hunk 7 above)
Comment (kichalla) [inline samples/ErrorHandlerSample/Startup.cs]: :+1:

**Comment by davidfowl** [comment]
@kichalla whats the deal with this PR

**Comment by kichalla** [comment]
@davidfowl this PR is about renaming the extension methods UseErrorHandler to UseExceptionHandler and UseErrorPage to UseDeveloperExceptionPage.

**Comment by davidfowl** [comment]
UseDeveloperExceptionPage lol :smile:

File: src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs (hunk 3 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageExtensions.cs]: Yay :smile:

**Comment by Eilon** [comment]
@davidfowl what's wrong with the name? It's accurate and relevant. And scares people away from using it in non-developer scenarios, which was a key motivator for this rename.

:shipit: from me.

**Comment by kichalla** [comment]
c46615dc534b55916e3c63bc9b0db89699b7c921

**Pull request closed without merging**